holoplot/__init__.py

```diff
@@ -0,0 +1,385 @@
+import holoviews
+from holoviews import (
+    extension as _extension, Dimensioned as _Dimensioned, HoloMap as _HoloMap
+)
+
+from bokeh.io import export_png as _export_png, show as _show, save as _save
+from bokeh.resources import CDN as _CDN
+
+from .converter import HoloViewsConverter
+
+renderer = holoviews.renderer('bokeh')
+
+# Register plotting interfaces
+def _patch_plot(self):
+    return HoloPlot(self)
+
+
+def patch(library, extension=None):
+    """
+    Patch library to support HoloViews based plotting API.
+    """
+    if not isinstance(library, list): library = [library]
+    if 'streamz' in library:
+        try:
+            import streamz.dataframe as sdf
+        except ImportError:
+            raise ImportError('Could not patch plotting API onto streamz. '
+                              'Streamz could not be imported.')
+        sdf.DataFrame.plot = property(_patch_plot)
+        sdf.DataFrames.plot = property(_patch_plot)
+        sdf.Series.plot = property(_patch_plot)
+        sdf.Seriess.plot = property(_patch_plot)
+    if 'pandas' in library:
+        try:
+            import pandas as pd
+        except:
+            raise ImportError('Could not patch plotting API onto pandas. '
+                              'Pandas could not be imported.')
+        pd.DataFrame.plot = property(_patch_plot)
+        pd.Series.plot = property(_patch_plot)
+    if 'dask' in library:
+        try:
+            import dask.dataframe as dd
+        except:
+            raise ImportError('Could not patch plotting API onto dask. '
+                              'Dask could not be imported.')
+        dd.DataFrame.plot = property(_patch_plot)
+        dd.Series.plot = property(_patch_plot)
+    if 'intake' in library:
+        try:
+            from intake.source.base import DataSource
+        except ImportError:
+            raise ImportError('Could not patch plotting API onto intake. '
+                              'Intake could not be imported.')
+        DataSource.plot = property(_patch_plot)
+    if 'xarray' in library:
+        try:
+            from xarray import DataArray
+        except ImportError:
+            raise ImportError('Could not patch plotting API onto xarray. '
+                              'Xarray could not be imported.')
+        DataArray.plot = property(_patch_plot)
+    if extension:
+        _extension(extension)
```

Comment:
Does this mean that no extension will be loaded when doing `from holoplot import xarray`?  I can understand why one wouldn't, to let the extension be selectable, but I strongly feel like an extension should be loaded so that the import is all that's required before using .plot.  Just getting some plot out of this system should be as easy as we can possibly make it, right?

Maybe `holoplot.xarray` could have something like `patch('xarray','auto')`, where `auto` tries loading the extensions according to a fixed importing order (`bokeh`,`matplotlib`, presumably?)?

Reply:
(With 'bokeh' first in the list because I think most people will be interested in HoloPlot because of its ability to return an interactive plot, though of course matplotlib-based HoloViews plots also have additional capabilities beyond what the native .plot() interface will support.)

Reply:
>Does this mean that no extension will be loaded when doing from holoplot import xarray? I can understand why one wouldn't, to let the extension be selectable, but I strongly feel like an extension should be loaded so that the import is all that's required before using .plot. Just getting some plot out of this system should be as easy as we can possibly make it, right?

Originally I did make it load the extension automatically, however loading the extension interferes with usage outside the notebook, because it's what holoplot uses to tell whether you're in a notebook or not. In other words the functionality described and shown in the Commandline & Scripts section [here](https://pyviz.github.io/holoplot/user_guide/Display_and_Saving.html) wouldn't work.

Reply:
> Worth noting that I have made no effort at all to support matplotlib yet. It's likely totally broken because options like width and height are always set.

The options used in holoplot make a good candidate set of things that we would like to work with the same syntax across backends, so that would be something to try to do separately from holoplot (respect those plot options in matplotlib). I do realize that the sizing options are the most difficult to respect given the libraries' different sizing models, but they are also the most important...

> Originally I did make it load the extension automatically, however loading the extension interferes with usage outside the notebook, because it's what holoplot uses to tell whether you're in a notebook or not.

I don't get it; I thought we've for many months been able to load hv.extension() in or outside of a notebook, and have it do the right thing in both cases.  I think that's crucial for having a smooth user experience; any single case where we have to tell people to do something different in different circumstances cuts down our audience dramatically.  Can't we have hv.extension() leave some marker that indicates whether it's in a notebook or not?  This is a very important issue; the instructions really have to look simple and clean for this interface to be compelling to its intended audience.

Reply:
>I don't get it; I thought we've for many months been able to load hv.extension() in or outside of a notebook, and have it do the right thing in both cases

That's because we don't offer any function that aims to display an object in or outside the notebook. For ease of use I've introduced a ``holoplot.show`` function that will display a plot in the notebook if the extension is enabled, or open a browser window with a static or server backed plot otherwise. I'd be okay making ``show`` always open a new window by default though.

How about making the imports load the extension and leaving patch as it is? Most users will just need the import.

Reply:
If you'll be making a PR for that, I'll change the notebooks to describe this way of doing things, as I already have extensive (but local and tiny) edits to them.